viz/pkg/healthcheck/healthcheck.go

```diff
@@ -0,0 +1,215 @@
+			WithHintAnchor("l5d-viz-ns-exists").
+			Fatal().
+			WithCheck(func(ctx context.Context) error {
+				vizNs, err := hc.KubeAPIClient().GetNamespaceWithExtensionLabel(ctx, "linkerd-viz")
+				if err == nil {
+					hc.vizNamespace = vizNs.Name
+				}
+				return err
+			}))
+
+	checkers = append(checkers,
+		*healthcheck.NewChecker("linkerd-viz ClusterRoles exist").
+			WithHintAnchor("l5d-viz-cr-exists").
+			Fatal().
+			Warning().
+			WithCheck(func(ctx context.Context) error {
+				return healthcheck.CheckClusterRoles(ctx, hc.KubeAPIClient(), true, []string{fmt.Sprintf("linkerd-%s-prometheus", hc.vizNamespace), fmt.Sprintf("linkerd-%s-tap", hc.vizNamespace)}, "")
+			}))
+
+	checkers = append(checkers,
+		*healthcheck.NewChecker("linkerd-viz ClusterRoleBindings exist").
+			WithHintAnchor("l5d-viz-crb-exists").
+			Fatal().
+			Warning().
+			WithCheck(func(ctx context.Context) error {
+				return healthcheck.CheckClusterRoleBindings(ctx, hc.KubeAPIClient(), true, []string{fmt.Sprintf("linkerd-%s-prometheus", hc.vizNamespace), fmt.Sprintf("linkerd-%s-tap", hc.vizNamespace)}, "")
+			}))
+
+	checkers = append(checkers,
+		*healthcheck.NewChecker("linkerd-viz ConfigMaps exist").
+			WithHintAnchor("l5d-viz-cm-exists").
+			Fatal().
+			Warning().
+			WithCheck(func(ctx context.Context) error {
+				return healthcheck.CheckConfigMaps(ctx, hc.KubeAPIClient(), hc.vizNamespace, true, []string{"linkerd-prometheus-config", "linkerd-grafana-config"}, "")
+			}))
+
+	checkers = append(checkers,
+		*healthcheck.NewChecker("tap API server has valid cert").
+			WithHintAnchor("l5d-tap-cert-valid").
+			Fatal().
+			WithCheck(func(ctx context.Context) error {
+				anchors, err := fetchTapCaBundle(ctx, hc.KubeAPIClient())
+				if err != nil {
+					return err
+				}
+				cert, err := hc.FetchCredsFromSecret(ctx, hc.vizNamespace, tapTLSSecretName)
+				if kerrors.IsNotFound(err) {
+					cert, err = hc.FetchCredsFromOldSecret(ctx, hc.vizNamespace, tapOldTLSSecretName)
+				}
+				if err != nil {
+					return err
+				}
+
+				identityName := fmt.Sprintf("linkerd-tap.%s.svc", hc.vizNamespace)
+				return hc.CheckCertAndAnchors(cert, anchors, identityName)
+			}))
+
+	checkers = append(checkers,
+		*healthcheck.NewChecker("tap API server cert is valid for at least 60 days").
+			WithHintAnchor("l5d-webhook-cert-not-expiring-soon").
+			Warning().
+			WithCheck(func(ctx context.Context) error {
+				cert, err := hc.FetchCredsFromSecret(ctx, hc.vizNamespace, tapTLSSecretName)
+				if kerrors.IsNotFound(err) {
+					cert, err = hc.FetchCredsFromOldSecret(ctx, hc.vizNamespace, tapOldTLSSecretName)
+				}
+				if err != nil {
+					return err
+				}
+				return hc.CheckCertAndAnchorsExpiringSoon(cert)
+			}))
+
+	checkers = append(checkers,
+		*healthcheck.NewChecker("tap API service is running").
+			WithHintAnchor("l5d-tap-api").
+			Warning().
+			WithCheck(func(ctx context.Context) error {
+				return hc.CheckAPIService(ctx, linkerdTapAPIServiceName)
+			}))
```

Comment:
It may be worth adding a retry to this checker now. When running `linkerd viz check` immediately after install, it produces the following warning:

```
‼ tap API service is running
    MissingEndpoints: endpoints for service/linkerd-tap in "linkerd-viz" have no addresses with port name "apiserver"
    see https://linkerd.io/checks/#l5d-tap-api for hints
```


Once you run `check` again though and the tap service is up and running it's fine.

Reply:
Makes sense. I've just added that into my latest push.